### ISSUE TITLE
Visualization setting: Solid scale feature

### ISSUE THREAD

**Issue by ricaun** [jeremytammik/RevitLookup#251]
First really at all amazing implementation of the visualization. Great job @Nice3point, pretty interesting the way you scale the vertex in model. 👏

One thing I think is missing is a way to scale the Solid, only because you can scale Face, Mesh, Line... And is easier to visualize if you can scale in front of the original geometry.

![image](https://github.com/jeremytammik/RevitLookup/assets/12437519/c3e9c240-951d-42c0-ac27-d86295d55506)

**Comment by Nice3point** [comment]
Thank you) makes sense. I didn't want to delay the first release as scaling Solid would take longer, that's the only reason why I added Cage instead of scale. Wait for the future, I'll add it

**Comment by Nice3point** [comment]
TODO:

We need to split volumes for tesselation, instead of using original solid. Sometimes faces array is empty (curtain walls)
![изображение](https://github.com/jeremytammik/RevitLookup/assets/20504884/84e6078b-4bf4-4551-98db-de52f59e1843)

**Comment by ricaun** [comment]
Interesting, I didn't know that.

This is the code for the Solid right?https://github.com/jeremytammik/RevitLookup/blob/dev/source/RevitLookup/Core/Visualization/SolidVisualizationServer.cs

**Comment by Nice3point** [comment]
@ricaun 

I added a minimum extrusion so that the rendering does not overlap the current geometry
![изображение](https://github.com/jeremytammik/RevitLookup/assets/20504884/65b41352-53ed-4a51-89f6-0c690cd2dfe5)

I need your opinion, to make the scaling a percentage, or to a specific value like Faces? In favour of an inch value goes the scaling of too small details, which even at x2 is hard to see
![изображение](https://github.com/jeremytammik/RevitLookup/assets/20504884/fd9e1870-6ecd-4d59-a206-bbbf8da341b4)

Scaling is performed from the Solid centre

**Comment by Nice3point** [comment]
also solved the problem with rendering of super thin solids, for comparison how it is displayed using RevitAPI UiDocument.Selection.SetReferences and visualisation

SetReferences:
![изображение](https://github.com/jeremytammik/RevitLookup/assets/20504884/682a1200-1697-4240-89dd-e02242abe1a4)

Visualization:
![изображение](https://github.com/jeremytammik/RevitLookup/assets/20504884/b0554ffb-96d5-449d-806e-0c06a2598e47)

**Comment by ricaun** [comment]
Neat!

`ComputeCentroid` is the center of gravity in the solid, I use the Origin from the BoundingBox to get the middle of the solid to perforce a scale in the middle of the solid. Could use the Min/Max as well.

```C#
/// <summary>
/// GetOrigin using bounding box transform origin
/// </summary>
/// <param name="solid"></param>
/// <returns></returns>
public static XYZ GetOrigin(this Solid solid)
{
    return solid.GetBoundingBox().Transform.Origin;
}
```
> I need your opinion, to make the scaling a percentage, or to a specific value like Faces? In favour of an inch value goes the scaling of too small details, which even at x2 is hard to see

I guess in the solid would make more sense to be percentage.

**Comment by Nice3point** [comment]
> I use the Origin from the BoundingBox

The behaviour is the same using Centroid and GetBoundingBox().Transform.Origin

> I guess in the solid would make more sense to be percentage.

I can try it, I'll have to test it for a couple of days to see what's more convenient

**Comment by ricaun** [comment]
> The behaviour is the same using Centroid and GetBoundingBox().Transform.Origin

Is not the same, if is a irregular solid the centroid gonna be different from the center of the boundbox.

Like this Cone with 1x1x1: 
* Blue: Origin
* Red: Centroid

![image](https://github.com/jeremytammik/RevitLookup/assets/12437519/408d3a0b-6911-495b-ab33-5c3ad0ee0fce)

> We need to determine what min and max values of scale are optimal and whether we should use a downscaling, probably not.

Maybe 1x to 10x, don't know if more than 10x is necessary,

**Comment by Nice3point** [comment]
![изображение](https://github.com/jeremytammik/RevitLookup/assets/20504884/df170fb7-6a14-4687-8562-98ff6ba3be5a)
Scaling by 4x this fireplace loads the cpu a lot and rendering takes 10 sec

**Comment by ricaun** [comment]
>  Scaling by 4x this fireplace loads the cpu a lot and rendering takes 10 sec

O.o
That is generating a lot of triangles per faces, the Triangulate in the face has a parameter, set to 0 just to test. The Mesh gonna be really bad.

**Comment by Nice3point** [comment]
> Triangulate in the face has a parameter, set to 0 just to test

![изображение](https://github.com/jeremytammik/RevitLookup/assets/20504884/7233d281-e9fd-4e77-b1bb-c9eaa397e3c4)
square for a round face?) no

**Comment by ricaun** [comment]
Hehehe, zero is too much. 🤣

Basically you can control the resolution of the triangulation, without the value Revit gonna define a value depending of the size of the face. If is a round face and is  really big, Revit gonna generate more triangles. 

If you set a value the quantity of triangles gonna be fixed maybe, no sure.
You could put a slide in the Face to see what the value does in the Triangulate. 
If you set to 0.1 probably gonna show a little better intend of a square 🤣

**Comment by Nice3point** [comment]
Released 2025.0.7. Tessellation kept as it is. In Release build it is not so slow. And slowdowns happen on very complex shaped objects, on normal objects it is extremely fast
